src/ledger/LedgerStateSnapshot.cpp

```diff
@@ -0,0 +1,296 @@
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerStateSnapshot.h"
+#include "bucket/BucketListSnapshot.h"
+#include "bucket/BucketManager.h"
+#include "bucket/BucketSnapshotManager.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "transactions/TransactionFrame.h"
+#include "transactions/TransactionUtils.h"
+#include "xdr/Stellar-ledger.h"
+
+namespace stellar
+{
+
+LedgerEntryWrapper::LedgerEntryWrapper(ConstLedgerTxnEntry&& entry)
+{
+    mEntry.emplace<ConstLedgerTxnEntry>(std::move(entry));
+}
+
+LedgerEntryWrapper::LedgerEntryWrapper(LedgerTxnEntry&& entry)
+{
+    mEntry.emplace<LedgerTxnEntry>(std::move(entry));
+}
+
+LedgerEntryWrapper::LedgerEntryWrapper(std::shared_ptr<LedgerEntry> entry)
+{
+    mEntry.emplace<std::shared_ptr<LedgerEntry>>(entry);
+}
+
+LedgerEntry const&
+LedgerEntryWrapper::current() const
+{
+    if (mEntry.index() == 0)
+    {
+        return std::get<0>(mEntry).current();
+    }
+    else if (mEntry.index() == 1)
+    {
+        return std::get<1>(mEntry).current();
+    }
+    else
+    {
+        return *std::get<2>(mEntry);
+    }
+}
+
+LedgerEntry&
+LedgerEntryWrapper::currentToModify()
+{
+    if (mEntry.index() == 0)
+    {
+        return std::get<0>(mEntry).current();
+    }
+    else if (mEntry.index() == 1)
+    {
+        throw std::runtime_error(
+            "LedgerEntryWrapper::current() called on a const LedgerEntry");
+    }
+    else
+    {
+        return *std::get<2>(mEntry);
+    }
+}
+
+LedgerEntryWrapper::operator bool() const
+{
+    if (mEntry.index() == 0)
+    {
+        return static_cast<bool>(std::get<0>(mEntry));
+    }
+    else if (mEntry.index() == 1)
+    {
+        return static_cast<bool>(std::get<1>(mEntry));
+    }
+    else
+    {
+        return static_cast<bool>(std::get<2>(mEntry));
+    }
+}
+
+LedgerHeaderWrapper::LedgerHeaderWrapper(LedgerHeader const& header)
+    : mHeader(std::make_shared<LedgerHeader>(header))
+{
+}
+
+LedgerHeaderWrapper::LedgerHeaderWrapper(LedgerTxnHeader&& header)
+    : mHeader(std::move(header))
+{
+}
+
+LedgerHeaderWrapper::LedgerHeaderWrapper(std::shared_ptr<LedgerHeader> header)
+    : mHeader(header)
+{
+}
+
+LedgerHeader&
+LedgerHeaderWrapper::currentToModify()
+{
+    if (mHeader.index() == 0)
+    {
+        return std::get<0>(mHeader).current();
+    }
+    else
+    {
+        return *std::get<1>(mHeader);
```

Comment:
I think we should throw here on the `shared_ptr<LedgerHeader>` case, given that you're not actually modifying the underlying ledger header like you would be if the base object is a `LedgerTxnHeader`

Reply:
It's a bit weird, but we need this for validation purposes where we modify ledgerSeq in LedgerHeader (e.g. https://github.com/stellar/stellar-core/blob/6cd475348c6a8372adea257d600efe9599ed8623/src/herder/TransactionQueue.cpp#L403)

I think this is ok though, since LedgerStateSnapshot never commits anything into the ledger, so we discard any changes when LedgerStateSnapshot goes out of scope.

Reply:
Hmm, I see. I don't really like that we rely on mutated state being "thrown out" in the tx queue path and that we mutate local copies sometimes in the shared_ptr path, and sometimes mutate commitable ledger headers with the expectation they won't get commited. This is probable a clean up for a later date though.